File: src/ResearchBridge.ts
export type ResearchOrigin = 'human' | 'betterboard' | 'engineering-lab' | 'openguin';
export type ResearchKind = 'observation' | 'measurement' | 'analysis' | 'annotation' | 'hypothesis' | 'suggestion' | 'warning' | 'decision';

export type ResearchBridgeEvent = {
  id: string;
  created_at_utc: string;
  origin: ResearchOrigin;
  kind: ResearchKind;
  text: string;
  refs?: string[];
};

export type ResearchContextForBridge = {
  question?: string;
  hypothesis?: string;
  notebook?: ResearchBridgeEvent[];
  annotations?: ResearchBridgeEvent[];
  lab_journey?: ResearchBridgeEvent[];
  engineering_results?: ResearchBridgeEvent[];
  ai_suggestions?: ResearchBridgeEvent[];
};

export type MeasurementSessionForBridge = {
  directory: string;
  created_at_utc: string;
  recipe_title: string;
  recipe_id?: string;
  sample_count: number;
  csv_path: string;
  metadata_path: string;
  physical_lab_csv_path: string;
  physical_lab_bridge_path: string;
  board_profile?: string;
  port?: string;
  acquisition_mode?: string;
};

export type ResearchBridgeV1 = {
  schema: 'betterboard.research-bridge/1.0';
  session_id: string;
  created_at_utc: string;
  producer: 'BetterBoard Studio';
  experiment: {
    title: string;
    question: string | null;
    hypothesis: string | null;
  };
  hardware: {
    board_profile: string | null;
    port: string | null;
  };
  evidence: {
    sample_count: number;
    data_csv: string;
    metadata_json: string;
    engineering_lab_compatibility_csv: string;
    legacy_bridge_json: string;
  };
  research_context: {
    notebook: ResearchBridgeEvent[];
    annotations: ResearchBridgeEvent[];
    lab_journey: ResearchBridgeEvent[];
  };
  engineering_lab: {
    imports: ResearchBridgeEvent[];
    results: ResearchBridgeEvent[];
    policy: string;
  };
  ai: {
    provider: 'OpenPenguin';
    suggestions: ResearchBridgeEvent[];
    policy: string;
  };
  provenance: ResearchBridgeEvent[];
};

export function researchSessionId(session: MeasurementSessionForBridge) {
  const tail = session.directory.split(/[\\/]/).filter(Boolean).pop();
  return tail || `session-${Date.parse(session.created_at_utc) || Date.now()}`;
}

export function buildResearchBridge(session: MeasurementSessionForBridge, context: ResearchContextForBridge = {}): ResearchBridgeV1 {
  const created = new Date().toISOString();
  const id = researchSessionId(session);
  const notebook = context.notebook ?? [];
  const annotations = context.annotations ?? [];
  const journey = context.lab_journey ?? [];
  const engineeringResults = context.engineering_results ?? [];
  const aiSuggestions = context.ai_suggestions ?? [];
  const contextualEvents = [...notebook, ...annotations, ...journey, ...engineeringResults, ...aiSuggestions];
  return {
    schema: 'betterboard.research-bridge/1.0',
    session_id: id,
    created_at_utc: created,
    producer: 'BetterBoard Studio',
    experiment: {
      title: session.recipe_title,
      question: context.question?.trim() || null,
      hypothesis: context.hypothesis?.trim() || null,
    },
    hardware: {
      board_profile: session.board_profile || null,
      port: session.port || null,
    },
    evidence: {
      sample_count: session.sample_count,
      data_csv: session.csv_path,
      metadata_json: session.metadata_path,
      engineering_lab_compatibility_csv: session.physical_lab_csv_path,
      legacy_bridge_json: session.physical_lab_bridge_path,
    },
    research_context: {
      notebook,
      annotations,
      lab_journey: journey,
    },
    engineering_lab: {
      imports: [],
      results: engineeringResults,
      policy: 'Engineering Lab may append derived analysis but must not replace BetterBoard raw measurement evidence.',
    },
    ai: {
      provider: 'OpenPenguin',
      suggestions: aiSuggestions,
      policy: 'OpenPenguin output is advisory and must remain distinguishable from measurements, calibration, and validated engineering results.',
    },
    provenance: [{
      id: `${id}:measurement-package`,
      created_at_utc: session.created_at_utc,
      origin: 'betterboard',
      kind: 'measurement',
      text: `${session.recipe_title} measurement package · ${session.sample_count} samples`,
      refs: [session.csv_path, session.metadata_path],
    }, ...contextualEvents],
  };
}

export function bridgeForOpenPenguin(bridge: ResearchBridgeV1) {
  return JSON.stringify({
    schema: bridge.schema,
    session_id: bridge.session_id,
    experiment: bridge.experiment,
    hardware: bridge.hardware,
    evidence: bridge.evidence,
    research_context: bridge.research_context,
    engineering_lab: bridge.engineering_lab,
    provenance: bridge.provenance,
    instruction: 'Keep raw measurement, human notes, Engineering Lab derived results, and AI suggestions distinct. Do not claim calibration or validation that is not present in the evidence.',
  }, null, 2);
}

File: src-tauri/src/research_bridge.rs
use chrono::Utc;
use serde::Serialize;
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::{fs, path::{Path, PathBuf}};

#[derive(Debug, Serialize)]
pub struct ResearchBridgeResult {
    path: String,
    session_id: String,
    schema: String,
}

fn measurement_root() -> PathBuf {
    if let Ok(home) = std::env::var("HOME") {
        return PathBuf::from(home).join("Documents").join("BetterBoard").join("measurements");
    }
    std::env::temp_dir().join("BetterBoard").join("measurements")
}

fn validated_session_dir(directory: &str) -> Result<PathBuf, String> {
    let root = measurement_root();
    fs::create_dir_all(&root).map_err(|e| e.to_string())?;
    let root = fs::canonicalize(&root).map_err(|e| e.to_string())?;
    let session = fs::canonicalize(directory).map_err(|e| format!("Measurement session does not exist: {e}"))?;
    if !session.is_dir() || !session.starts_with(&root) {
        return Err("Research bridge creation is restricted to BetterBoard measurement sessions.".into());
    }
    Ok(session)
}

fn sha256_file(path: &Path) -> Result<String, String> {
    let bytes = fs::read(path).map_err(|e| format!("Could not read {}: {e}", path.display()))?;
    let mut hasher = Sha256::new();
    hasher.update(bytes);
    Ok(format!("{:x}", hasher.finalize()))
}

fn file_record(dir: &Path, name: &str, role: &str) -> Value {
    let path = dir.join(name);
    if !path.is_file() {
        return json!({"role": role, "name": name, "present": false});
    }
    json!({
        "role": role,
        "name": name,
        "present": true,
        "sha256": sha256_file(&path).ok(),
        "bytes": fs::metadata(&path).ok().map(|m| m.len()),
    })
}

fn build_bridge(directory: &str) -> Result<(PathBuf, Value), String> {
    let dir = validated_session_dir(directory)?;
    let metadata_path = dir.join("metadata.json");
    let metadata_text = fs::read_to_string(&metadata_path)
        .map_err(|e| format!("Could not read measurement metadata: {e}"))?;
    let metadata: Value = serde_json::from_str(&metadata_text)
        .map_err(|e| format!("Measurement metadata is invalid JSON: {e}"))?;
    let session_id = dir.file_name().and_then(|v| v.to_str()).unwrap_or("measurement-session").to_string();
    let created_at = Utc::now().to_rfc3339();

    let bridge = json!({
        "schema": "betterboard.research-bridge/1.0",
        "session_id": session_id,
        "created_at_utc": created_at,
        "producer": "BetterBoard Studio",
        "experiment": {
            "title": metadata.get("recipe_title").cloned().unwrap_or(Value::Null),
            "recipe_id": metadata.get("recipe_id").cloned().unwrap_or(Value::Null),
            "question": Value::Null,
            "hypothesis": Value::Null
        },
        "hardware": {
            "board_profile": metadata.get("board_profile").cloned().unwrap_or(Value::Null),
            "port": metadata.get("port").cloned().unwrap_or(Value::Null),
            "baud": metadata.get("baud").cloned().unwrap_or(Value::Null),
            "identity_boundary": "Board profile and serial transport are provenance, not proof of exact physical board revision or electrical limits."
        },
        "firmware": {
            "sha256": metadata.get("firmware_sha256").cloned().unwrap_or(Value::Null),
            "recipe_parameters": metadata.get("recipe_parameters").cloned().unwrap_or(json!({}))
        },
        "evidence": {
            "sample_count": metadata.get("sample_count").cloned().unwrap_or(Value::Null),
            "columns": metadata.get("columns").cloned().unwrap_or(json!([])),
            "units": metadata.get("units").cloned().unwrap_or(json!([])),
            "primary_column": metadata.get("primary_column").cloned().unwrap_or(Value::Null),
            "sample_rate_hz": metadata.get("sample_rate_hz").cloned().unwrap_or(Value::Null),
            "files": [
                file_record(&dir, "data.csv", "raw-measurement"),
                file_record(&dir, "metadata.json", "measurement-metadata"),
                file_record(&dir, "physical_lab_v1.csv", "engineering-lab-compatibility"),
                file_record(&dir, "physical_lab_bridge.json", "legacy-engineering-lab-bridge")
            ]
        },
        "research_context": {
            "notebook": [],
            "annotations": [],
            "lab_journey": []
        },
        "engineering_lab": {
            "imports": [],
            "results": [],
            "policy": "Derived analysis must remain separate from raw BetterBoard measurement evidence."
        },
        "ai": {
            "provider": "OpenPenguin",
            "suggestions": [],
            "policy": "AI output is advisory context and must not be represented as measurement, calibration, or validated engineering result."
        },
        "provenance": [{
            "origin": "betterboard",
            "kind": "measurement-package",
            "created_at_utc": created_at,
            "source": "metadata.json",
            "immutable_raw_evidence": true
        }],
        "scientific_boundary": metadata.get("scientific_boundary").cloned().unwrap_or(Value::Null)
    });
    Ok((dir, bridge))
}

#[tauri::command]
pub fn research_bridge_create(directory: String) -> Result<ResearchBridgeResult, String> {
    let (dir, bridge) = build_bridge(&directory)?;
    let path = dir.join("betterboard_bridge.json");
    fs::write(&path, serde_json::to_string_pretty(&bridge).map_err(|e| e.to_string())?)
        .map_err(|e| format!("Could not write research bridge: {e}"))?;
    Ok(ResearchBridgeResult {
        path: path.display().to_string(),
        session_id: bridge.get("session_id").and_then(Value::as_str).unwrap_or_default().to_string(),
        schema: "betterboard.research-bridge/1.0".into(),
    })
}

#[tauri::command]
pub fn research_bridge_context(directory: String) -> Result<String, String> {
    let (_, bridge) = build_bridge(&directory)?;
    let compact = json!({
        "schema": bridge.get("schema"),
        "session_id": bridge.get("session_id"),
        "experiment": bridge.get("experiment"),
        "hardware": bridge.get("hardware"),
        "firmware": bridge.get("firmware"),
        "evidence": bridge.get("evidence"),
        "research_context": bridge.get("research_context"),
        "engineering_lab": bridge.get("engineering_lab"),
        "scientific_boundary": bridge.get("scientific_boundary")
    });
    serde_json::to_string_pretty(&compact).map_err(|e| e.to_string())
}
